### PR TITLE
[DependencyInjection] Document %env(...)% dynamic parameters

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Ref. https://github.com/symfony/symfony/pull/19681

### CODE REVIEW

File: configuration/external_parameters.rst [hunk 1]
```diff
@@ -80,24 +123,51 @@ You can now reference these parameters wherever you need them.
...
+                <parameter key="env(DATABASE_HOST)">localhost</parameter>
+            </parameters>
+         </container>
+
+    .. code-block:: php
+
+        $container->setParameter('env(DATABASE_HOST)', 'localhost');
+
+Unlike ``SYMFONY__`` prefixed environment variables, ``%env(...)%`` parameters are
+truned into run-time environment variable reads, so that dumped containers can
```
Comment (tgalopin) [inline configuration/external_parameters.rst]: turned

File: configuration/external_parameters.rst [hunk 2]
```diff
@@ -80,24 +123,51 @@ You can now reference these parameters wherever you need them.
...
 
     .. code-block:: php
 
         $container->loadFromExtension('doctrine', array(
             'dbal' => array(
-                'driver'   => 'pdo_mysql',
-                'dbname'   => 'symfony_project',
-                'user'     => '%database.user%',
-                'password' => '%database.password%',
+                'host' => '%env(DATABASE_HOST)%',
```
Comment (lyrixx) [inline configuration/external_parameters.rst]: One can wonder "why not using `getenv` directly". **I** know why, but I could be nice to explain why ?!

Comment (nicolas-grekas) [inline configuration/external_parameters.rst]: Not sure, at least I don't know how to phrase it :) This is explained indirectly a few lines below this one. If someone has any suggestions, please do, otherwise I might leave it asis.

Comment (wouterj) [inline configuration/external_parameters.rst]: I propose the following:

 * Keep the "Environment variables" section title and document the `%env(...)%` syntax in there. Explain what by reference means ("You can reference environment variables using the `%env(...)%` syntax. The actual value will be resolved when using the container: [...]").
 * Add a small `.. note::` directive documenting `SYMFONY__` (I personally don't see any reason to use this, other than BC. In other words, why not deprecate it?). This will say something like: "You can also directly set parameters by prefixing the environment variable with `SYFMONY__` (e.g. `SYMFONY__DATABASE__USER` would set a `database.user` parameter). This value will be resolved when warming up the cache and will not change at runtime."

File: configuration/external_parameters.rst [hunk 3]
```diff
@@ -80,24 +123,51 @@ You can now reference these parameters wherever you need them.
...
+    .. code-block:: php
+
+        $container->setParameter('env(DATABASE_HOST)', 'localhost');
+
+Unlike ``SYMFONY__`` prefixed environment variables, ``env(...)`` parameters are
+turned into run-time environment variable reads, so that dumped containers can
+be reconfigured dynamically even after being compiled.
+
+.. versionadded:: 3.2
+    ``env(...)`` parameters were introduced in Symfony 3.2.
```
Comment (wouterj) [inline configuration/external_parameters.rst]: please put the versionadded directive after the section header, instead of at the end

**Comment by wouterj** [comment]
status: needs work

File: configuration/external_parameters.rst (hunk 2 above)
Comment (nicolas-grekas) [inline configuration/external_parameters.rst]: @wouterj PR updated accordingly

File: configuration/external_parameters.rst [hunk 4]
```diff
@@ -80,24 +40,84 @@ You can now reference these parameters wherever you need them.
...
         ));
 
+``env(...)`` parameters can have default values be set and used whenever their
+corresponding environment variables are not found:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        parameters:
```
Comment (weaverryan) [inline configuration/external_parameters.rst]: Add `#app/config/parameters.yml` above this - it's *probably* the file where they will define the default value

File: configuration/external_parameters.rst [hunk 5]
```diff
@@ -80,24 +40,84 @@ You can now reference these parameters wherever you need them.
...
-                'driver'   => 'pdo_mysql',
-                'dbname'   => 'symfony_project',
-                'user'     => '%database.user%',
-                'password' => '%database.password%',
+                'host' => '%env(DATABASE_HOST)%',
             )
         ));
 
+``env(...)`` parameters can have default values be set and used whenever their
+corresponding environment variables are not found:
```
Comment (weaverryan) [inline configuration/external_parameters.rst]:
```rst
You can also give the ``env(...)`` parameters a default value: the default value will be used whenever
the corresponding environment variable is *not* found:
```

File: configuration/external_parameters.rst [hunk 6]
```diff
@@ -80,24 +40,84 @@ You can now reference these parameters wherever you need them.
...
+        <Directory "/path/to/symfony_2_app/web">
+            AllowOverride All
+            Allow from All
+        </Directory>
+    </VirtualHost>
+
+.. note::
+
+    All popular web servers support setting environment variables. Read their
+    documentation to know how.
```
Comment (weaverryan) [inline configuration/external_parameters.rst]: ... to find out how.

File: configuration/external_parameters.rst [hunk 7]
```diff
@@ -80,24 +40,84 @@ You can now reference these parameters wherever you need them.
...
+
+    .. code-block:: bash
+
+        $ export DATABASE_USER=user
+        $ export DATABASE_PASSWORD=secret
+
+.. note::
+
+    You can also define the default value of any existing parameters using
+    special environement variables named after their corresponding parameter
```
Comment (weaverryan) [inline configuration/external_parameters.rst]: environment

**Review by weaverryan — CHANGES_REQUESTED** [review]
Status: Reviewed

Setting this as reviewed because all of my comments are very minor!

**Comment by weaverryan** [comment]
Status: Reviewed

Just needs a few minor changes.

File: configuration/external_parameters.rst [hunk 8]
```diff
@@ -80,24 +40,84 @@ You can now reference these parameters wherever you need them.
...
+    prefixed with ``SYMFONY__`` after replacing dots by double underscores
+    (e.g. ``SYMFONY__FOO__BAR`` to set the default value of the ``foo.bar``
```
Comment (wouterj) [inline configuration/external_parameters.rst]: let's keep using a real example here: `(e.g. `SYMFONY__DATABASE__USER` to set the default value of the `database.user` parameter).

**Comment by nicolas-grekas** [comment]
Comments addressed, thanks for the review.

File: configuration/external_parameters.rst (hunk 2 above)
Comment (javiereguiluz) [inline configuration/external_parameters.rst]: @wouterj I agree with your deprecation proposal. I've created https://github.com/symfony/symfony/issues/20089 for that.

**Comment by wouterj** [comment]
👍

**Review by wouterj — APPROVED** [review]

File: configuration/external_parameters.rst [hunk 9]
```diff
@@ -80,24 +40,86 @@ You can now reference these parameters wherever you need them.
...
+will be used whenever the corresponding environment variable is *not* found:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        #app/config/parameters.yml
+
+        parameters:
+            env(DATABASE_HOST): localhost
```
Comment (stof) [inline configuration/external_parameters.rst]: @mickadoo your code is **not** what is documented here

Comment (mickadoo) [inline configuration/external_parameters.rst]: @stof I just thought I'd try it out, didn't work though

Comment (wouterj) [inline configuration/external_parameters.rst]: What's documented here is:
```yaml
parameters:
    database_driver: '%env(DATABASE_DRIVER)%'
    env(DATABASE_DRIVER): pdo_mysql
```

Which, as you can see, is quite different from the format you tested. If there is something we can do to improve the description of the feature, please tell us!

Comment (mickadoo) [inline configuration/external_parameters.rst]: @wouterj Ah, I get it now. First I thought it was maybe a mistake in the docs cos the syntax is different and requires people to change their parameter names. Sorry about that!

File: configuration/external_parameters.rst [hunk 10]
```diff
@@ -13,65 +13,25 @@ you to easily do this.
...
-applied to the resulting parameter name:
+.. versionadded:: 3.2
+    ``env(...)`` parameters were introduced in Symfony 3.2.
 
-* ``SYMFONY__`` prefix is removed;
-* Parameter name is lowercased;
-* Double underscores are replaced with a period, as a period is not
-  a valid character in an environment variable name.
+You can reference environment variables by using special parameters named after
+the variables you want to use encapsed between ``env(...)``. Their actual values
```
Comment (xabbuh) [inline configuration/external_parameters.rst]: "enclosed" or "encapsulated" instead of "encapsed"?

File: configuration/external_parameters.rst [hunk 11]
```diff
@@ -80,24 +40,86 @@ You can now reference these parameters wherever you need them.
...
 
+You can also give the ``env(...)`` parameters a default value: the default value
+will be used whenever the corresponding environment variable is *not* found:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        #app/config/parameters.yml
+
```
Comment (xabbuh) [inline configuration/external_parameters.rst]: blank line should be removed

File: configuration/external_parameters.rst [hunk 12]
```diff
@@ -80,24 +40,86 @@ You can now reference these parameters wherever you need them.
...
+        parameters:
+            env(DATABASE_HOST): localhost
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
```
Comment (xabbuh) [inline configuration/external_parameters.rst]: same for the PHP config example

File: configuration/external_parameters.rst [hunk 13]
```diff
@@ -80,24 +40,86 @@ You can now reference these parameters wherever you need them.
...
+            <parameters>
+                <parameter key="env(DATABASE_HOST)">localhost</parameter>
+            </parameters>
+         </container>
+
+    .. code-block:: php
+
+        $container->setParameter('env(DATABASE_HOST)', 'localhost');
+
+If you're using Apache, environment variables can be set e.g. using the following
```
Comment (xabbuh) [inline configuration/external_parameters.rst]: Everywhere else we give examples for Nginx too IIRC. I think we should do the same here.

File: configuration/external_parameters.rst (hunk 9 above)
Comment (wouterj) [inline configuration/external_parameters.rst]: No problem @mickadoo thanks for test-casing this documentation proposal.

I think we should include the  `database_driver: '%env(DATABASE_DRIVER)%'` line in this example to avoid confusing for other readers.

**Comment by wouterj** [comment]
status: needs work

ping @nicolas-grekas

File: configuration/external_parameters.rst (hunk 12 above)
Comment (nicolas-grekas) [inline configuration/external_parameters.rst]: Do you have an example of such filenames for xml & php?

**Comment by nicolas-grekas** [comment]
comments addressed
Status: needs review

File: configuration/external_parameters.rst (hunk 12 above)
Comment (xabbuh) [inline configuration/external_parameters.rst]: We usually simply replace the file extension. So here we would have `app/config/parameters.yml`, `app/config/parameters.xml`, and `app/config/parameters.php`. By the way, can you also add similar comments in the example above to, but using `config.yml`, `config.xml`, and `config.php` there?

Comment (xabbuh) [inline configuration/external_parameters.rst]: Ah well, the previous example is not new. We should add the comments there in older branches too and can probably better do that in a separate PR.

File: configuration/external_parameters.rst [hunk 14]
```diff
@@ -13,93 +13,25 @@ you to easily do this.
...
-    variables.
-
-    Also, in order for your console to work (which does not use a web server),
-    you must export these as shell variables. On a Unix system, you can run
-    the following:
-
-    .. code-block:: terminal
+You can reference environment variables by using special parameters named after
+the variables you want to use enclosed between ``env()``. Their actual values
+will be resolved at runtime, so that dumped containers can be reconfigured
```
Comment (javiereguiluz) [inline configuration/external_parameters.rst]: A quick question about this:

```
Their actual values will be resolved at runtime [...]
```

Are they resolved at runtime *every time* they are used ... or are they resolved *just* the first time they are used and then Symfony reuses the cached value? I say this because of [the code of the getEnv() method](https://github.com/symfony/symfony/pull/19681/files#diff-1df5b32be44d0f592ec285c6bbf386bbR386).

Comment (xabbuh) [inline configuration/external_parameters.rst]: Per request they are resolved once. But if you change the value, during the next request you will see the new value (again resolved only once during this request). Should we make that more clear somehow?

Comment (javiereguiluz) [inline configuration/external_parameters.rst]: Thanks @xabbuh. I wouldn't add any notice about this because that behavior is exactly what we should expect.

Comment (nicolas-grekas) [inline configuration/external_parameters.rst]: I added `(once per request)`, good enough?

File: configuration/external_parameters.rst [hunk 15]
```diff
@@ -13,119 +13,114 @@ you to easily do this.
...
+        $container->setParameter('database_host', '%env(DATABASE_HOST)%');
+        $container->setParameter('env(DATABASE_HOST)', 'localhost');
+
+Setting environment variables is generally done at the web server level or in the
+terminal. If you're using Apache, Nginx or just the console, you can use e.g. one
+of the following:
+
+.. configuration-block::
+
+    .. code-block:: apache
```
Comment (javiereguiluz) [inline configuration/external_parameters.rst]: @wouterj thanks for the heads up. It will be merged in symfony.com soon:

apache-sphinx

Comment (wouterj) [inline configuration/external_parameters.rst]: @javiereguiluz please note that you also have to add `'nginx': 'NGINX'`

Comment (javiereguiluz) [inline configuration/external_parameters.rst]: @wouterj thanks! It's fixed now. I opted for `Nginx` as the human-readable name instead of `NGINX`. The official website mixes `nginx`, `Nginx` and `NGINX` ... but the Wikipedia always uses `Nginx`.

**Comment by wouterj** [comment]
👍 
status: reviewed

Note to merger: Please fix the `conf.py` configuration to suppor the new configuration block formats.

**Comment by wouterj** [comment]
ping @javiereguiluz @weaverryan @xabbuh please review & vote

**Review by javiereguiluz — APPROVED** [review]
👍

**Comment by xabbuh** [comment]
👍

**Comment by xabbuh** [comment]
Thank you Nicolas.
